### PR TITLE
Feature/gripper skeleton

### PR DESCRIPTION
**Pull request by sachifig**
Gripper subsystem skeleton contains GrabDiscCommand, ReleaseDiscCommand, test files for each command, and subsystem class.

### CODE REVIEW

File: Competition/src/main/java/competition/subsystems/gripper/GripperSubsystem.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Inject
+    public GripperSubsystem(CommonLibFactory clf, ElectricalContract2019 contract) {
+        gripperReady = contract.isGripperReady();
+
+        if (gripperReady) {
+            gripperPiston = clf.createSolenoid(contract.getGripperSolenoid().channel);
+            gripperPiston.setInverted(contract.getGripperSolenoid().inverted);
+        
+            diskSensor = clf.createDigitalInput(-1);
```
Comment (JohnGilb) [inline Competition/src/main/java/competition/subsystems/gripper/GripperSubsystem.java]: You should get this from the contract. Also, do the same setInverted call as above.

File: Competition/src/main/java/competition/subsystems/gripper/GripperSubsystem.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        }
+        
+    }
+
+    public void grabHatch(){
+        if (gripperReady)
+        {
+            gripperPiston.setOn(true);
+        }
+
```
Comment (JohnGilb) [inline Competition/src/main/java/competition/subsystems/gripper/GripperSubsystem.java]: remove this blank line

File: Competition/src/main/java/competition/subsystems/gripper/GripperSubsystem.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            gripperPiston.setOn(true);
+        }
+        
+    }
+    public void releaseHatch(){
+        if (gripperReady)
+        {
+            gripperPiston.setOn(false);
+        }
+
```
Comment (JohnGilb) [inline Competition/src/main/java/competition/subsystems/gripper/GripperSubsystem.java]: remove this blank line

File: Competition/src/main/java/competition/subsystems/gripper/GripperSubsystem.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public boolean currentlyHasDisk()
+    {
+        if (gripperReady)
+        {
+            return diskSensor.get();
+        }
+        return false;
+
```
Comment (JohnGilb) [inline Competition/src/main/java/competition/subsystems/gripper/GripperSubsystem.java]: remove this blank line

File: Competition/src/main/java/competition/subsystems/gripper/commands/GrabDiscCommand.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+import competition.subsystems.gripper.GripperSubsystem;
+import xbot.common.command.BaseCommand;
+
+public class GrabDiscCommand extends BaseCommand{
+    final GripperSubsystem gripperSubsystem;
+
+    @Inject
+    public GrabDiscCommand(GripperSubsystem gripper) {
+        gripperSubsystem=gripper;
```
Comment (JohnGilb) [inline Competition/src/main/java/competition/subsystems/gripper/commands/GrabDiscCommand.java]: add some spaces, so `gripperSubsystem = gripper;`

File: Competition/src/main/java/competition/subsystems/gripper/commands/ReleaseDiscCommand.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+import competition.subsystems.gripper.GripperSubsystem;
+import xbot.common.command.BaseCommand;
+
+public class ReleaseDiscCommand extends BaseCommand{
+    final GripperSubsystem gripperSubsystem;
+
+    @Inject
+    public ReleaseDiscCommand(GripperSubsystem gripper) {
+        gripperSubsystem=gripper;
```
Comment (JohnGilb) [inline Competition/src/main/java/competition/subsystems/gripper/commands/ReleaseDiscCommand.java]: Same as above

File: Competition/src/test/java/competition/subsystems/gripper/GrabDiscCommandTest.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        grabDisc = this.injector.getInstance(GrabDiscCommand.class);
+        gripperSubsystem = this.injector.getInstance(GripperSubsystem.class);
+    }
+
+    @Test
+    public void testGrabDisc() {
+        gripperSubsystem.gripperPiston.setOn(false);
+        assertFalse("grabHatchet is false", gripperSubsystem.gripperPiston.getAdjusted());
+
+        gripperSubsystem.grabHatch();
```
Comment (JohnGilb) [inline Competition/src/test/java/competition/subsystems/gripper/GrabDiscCommandTest.java]: This method doesn't actually use the GrabDiscCommand. You should call `GrabDiscCommand.initialize()`

File: Competition/src/test/java/competition/subsystems/gripper/GripperSubsystemTest.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        ((MockDigitalInput)gripperSubsystem.diskSensor).setValue(true);
+        // Check the method returns true 
+        assertTrue("Sensor does have disk", gripperSubsystem.currentlyHasDisk());;       
+    }
+
+    @Test
+    public void testGrabHatch(){
+        GripperSubsystem gripperSubsystem = this.injector.getInstance(GripperSubsystem.class);
+        gripperSubsystem.gripperPiston.setOn(false);
+        gripperSubsystem.grabHatch();
```
Comment (JohnGilb) [inline Competition/src/test/java/competition/subsystems/gripper/GripperSubsystemTest.java:39]: before this line, check to make sure the gripper is set off.

File: Competition/src/test/java/competition/subsystems/gripper/GripperSubsystemTest.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        assertTrue("Piston is true", (gripperSubsystem.gripperPiston.getAdjusted()));
+    }
+
+    @Test
+    public void testreleaseHatch(){
+        GripperSubsystem gripperSubsystem = this.injector.getInstance(GripperSubsystem.class);
+        gripperSubsystem.gripperPiston.setOn(true);
+        gripperSubsystem.releaseHatch();
```
Comment (JohnGilb) [inline Competition/src/test/java/competition/subsystems/gripper/GripperSubsystemTest.java:48]: before this line, check to make sure the gripper is set on.

File: Competition/src/test/java/competition/subsystems/gripper/ReleaseDiscCommandTest.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        grabDisc = this.injector.getInstance(GrabDiscCommand.class);
+        gripperSubsystem = this.injector.getInstance(GripperSubsystem.class);
+    }
+
+    @Test
+    public void testGrabDisc() {
+        gripperSubsystem.gripperPiston.setOn(true);
+        assertTrue("grabHatchet is true", gripperSubsystem.gripperPiston.getAdjusted());
+
+        gripperSubsystem.releaseHatch();
```
Comment (JohnGilb) [inline Competition/src/test/java/competition/subsystems/gripper/ReleaseDiscCommandTest.java]: You need to use the Command, not the Subsystem (see my other comments on the other command test)

**Review by JohnGilb — CHANGES_REQUESTED** [review]

**Comment by sachifig** [comment]
@JohnGilb We finished editing.

**Pull request merged**